### ISSUE TITLE
Please add details of 'correct' nvme tool output to README

### ISSUE THREAD

**Issue by joel-wright** [damige/linux-nvme#11]
The README currently states that a user can run "nvme get-feature -f 0x0c -H /dev/nvme0" to find out if the patch is working, but no details of what the expected output should be in this case.

Should we just expect to see that APST is enabled?
Should the table of states printed include specific things?

The reason I ask this is that the output for me says that APST is enabled, but the table entries are all the same and do not seem to indicate any power saving states being available to transition into.

**Comment by damige** [comment]
My output is:

get-feature:0xc (Autonomous Power State Transition), Current value:0x000001
	Autonomous Power State Transition Enable (APSTE): Enabled
	Auto PST Entries	.................
	Entry[ 0]   
	.................
	Idle Time Prior to Transition (ITPT): 275 ms
	Idle Transition Power State   (ITPS): 3
	.................
	Entry[ 1]   
	.................
	Idle Time Prior to Transition (ITPT): 275 ms
	Idle Transition Power State   (ITPS): 3
	.................
	Entry[ 2]   
	.................
	Idle Time Prior to Transition (ITPT): 275 ms
	Idle Transition Power State   (ITPS): 3
	.................
	Entry[ 3]   
	.................
	Idle Time Prior to Transition (ITPT): 1200 ms
	Idle Transition Power State   (ITPS): 4


Entry [4-31] are:
	Idle Time Prior to Transition (ITPT): 0 ms
	Idle Transition Power State   (ITPS): 0

My device (SM951) only has 4 power states. Some things to verify its working:

-       Autonomous Power State Transition Enable (APSTE): **Enabled**
-       "/sys/class/nvme/nvme0/power/pm_qos_latency_tolerance_us" exists 
-       Measurably lower power usage when ssd is idle

Regarding you not having any power states listed seems strange. Possibly verify if its working by my suggestions above. If they are not, doublecheck if your nvme ssd is APST capable.

**Comment by joel-wright** [comment]
Thanks for the reply, and thanks for the sample output. I think I have something strange going on here. After patching my kernel, my Toshiba drive shows `Autonomous Power State Transition Enable (APSTE): Enabled` and the `pm_qos_latency_tolerance_us` sysfs entry exists, but the `get-feature`  output is odd. My patched 4.9.x and 4.10.x kernels produce the following table:

```
get-feature:0x0c (Autonomous Power State Transition), Current value: 0x000001
        Autonomous Power State Transition Enable (APSTE): Enabled
        Auto PST Entries        .................
        Entry[ 0]   
        .................
        Idle Time Prior to Transition (ITPT): 0 ms
        Idle Transition Power State   (ITPS): 0
        .................
        Entry[ 1]   
        .................
        Idle Time Prior to Transition (ITPT): 0 ms
        Idle Transition Power State   (ITPS): 0
        .................
        Entry[ 2]   
        .................
        Idle Time Prior to Transition (ITPT): 0 ms
        Idle Transition Power State   (ITPS): 0
        .................
        Entry[ 3]   
        .................
        Idle Time Prior to Transition (ITPT): 0 ms
        Idle Transition Power State   (ITPS): 0
        .................
        Entry[ 4]   
        .................
        Idle Time Prior to Transition (ITPT): 0 ms
        Idle Transition Power State   (ITPS): 0
        .................
```

However, on an unpatched 4.8.x Ubuntu kernel `get-feature` gives me the following table (but obviously with APSTE disabled):

```
get-feature:0x0c (Autonomous Power State Transition), Current value: 00000000
        Autonomous Power State Transition Enable (APSTE): Disabled
        Auto PST Entries        .................
        Entry[ 0]   
        .................
        Idle Time Prior to Transition (ITPT): 50 ms
        Idle Transition Power State   (ITPS): 3
        .................
        Entry[ 1]   
        .................
        Idle Time Prior to Transition (ITPT): 50 ms
        Idle Transition Power State   (ITPS): 3
        .................
        Entry[ 2]   
        .................
        Idle Time Prior to Transition (ITPT): 50 ms
        Idle Transition Power State   (ITPS): 3
        .................
        Entry[ 3]   
        .................
        Idle Time Prior to Transition (ITPT): 6000 ms
        Idle Transition Power State   (ITPS): 4
        .................
        Entry[ 4]   
        .................
        Idle Time Prior to Transition (ITPT): 0 ms
        Idle Transition Power State   (ITPS): 4
        .................
```

So APSTE is being enabled on the patched kernels, but the tables contains no useful entries!

The output from `nvme id-ctrl /dev/nvme0` is the same for all the kernels:

```
NVME Identify Controller:
vid     : 0x1179
ssvid   : 0x1179
sn      : X6HS10IBT1CT        
mn      : THNSN51T02DUK NVMe TOSHIBA 1024GB       
fr      : 5KDA4101
rab     : 1
ieee    : 00080d
cmic    : 0
mdts    : 0
cntlid  : 0
ver     : 0
rtd3r   : 0
rtd3e   : 0
oaes    : 0
oacs    : 0x17
acl     : 3
aerl    : 3
frmw    : 0x2
lpa     : 0x2
elpe    : 127
npss    : 4
avscc   : 0
apsta   : 0x1
wctemp  : 351
cctemp  : 355
mtfa    : 0
hmpre   : 0
hmmin   : 0
tnvmcap : 0
unvmcap : 0
rpmbs   : 0
sqes    : 0x66
cqes    : 0x44
nn      : 1
oncs    : 0x1e
fuses   : 0
fna     : 0
vwc     : 0x1
awun    : 255
awupf   : 0
nvscc   : 0
acwu    : 0
sgls    : 0
ps    0 : mp:6.00W operational enlat:0 exlat:0 rrt:0 rrl:0
          rwt:0 rwl:0 idle_power:- active_power:-
ps    1 : mp:2.40W operational enlat:0 exlat:0 rrt:1 rrl:1
          rwt:1 rwl:1 idle_power:- active_power:-
ps    2 : mp:1.90W operational enlat:0 exlat:0 rrt:2 rrl:2
          rwt:2 rwl:2 idle_power:- active_power:-
ps    3 : mp:0.0120W non-operational enlat:5000 exlat:25000 rrt:3 rrl:3
          rwt:3 rwl:3 idle_power:- active_power:-
ps    4 : mp:0.0060W non-operational enlat:100000 exlat:80000 rrt:4 rrl:4
          rwt:4 rwl:4 idle_power:- active_power:-
```

So clearly there is some indication that power state transitions should be possible. I guess my only option is to hammer the drive and see if there's any noticeable difference in power usage, but that's hardly the most reliable way to work out if the power state transitions are working. I really wish there was an easy way to display the current power state of the drive...

**Comment by damige** [comment]
You don't need to hammer it. Just leave it idle on the desktop. You should see an improvement (use powerstat when running on battery) on idle power usage when APST is patched into the kernel.

I have not seen anyone use a Toshiba drive with these patches before. Like you say it does seem like it must be possible to use APST. At the very least it seems like a display bug or maybe the current version of the patch is incompatible with your device.  Either way its outside the realm of what i'm capable of fixing myself. To get it debugged you would have to contact Andy Lutomirski, as he made these patches.

Regarding displaying the current power state of the device, i don't know of a way. Though i'm pretty sure it already exists and i dont know if it, or if it doesn't it could be exposed to the user in a future version of the patchset.

Please keep us informed!

**Comment by joel-wright** [comment]
Thanks for taking the time to help with this. While I can't guarantee that *all* the differences in power usage between my unpatched 2.8.x kernel and my patched 2.9.x kernel are down to Andy's APST patch, the difference between them is huge (both runs below were immediately after reboot).

Unpatched 2.8.x:

```
Running for 300.0 seconds (30 samples at 10.0 second intervals).
Power measurements will start in 180 seconds time.

  Time    User  Nice   Sys  Idle    IO  Run Ctxt/s  IRQ/s Fork Exec Exit  Watts
20:17:10   0.7   0.0   0.4  98.9   0.0    1    858    294    3    2    3  19.51 
20:17:20   0.5   0.0   0.4  99.1   0.0    1    746    252    0    0    0  19.31 
20:17:30   0.5   0.0   0.4  99.1   0.0    1    779    176    2    0    2  19.33 
20:17:40   0.5   0.0   0.3  99.2   0.0    1    708    139    0    0    0  19.53 
20:17:50   0.6   0.0   0.5  98.9   0.0    2    796    252    0    0    0  19.27 
20:18:00   0.5   0.0   0.3  99.2   0.0    1    734    224    0    0    0  19.40 
20:18:10   0.5   0.0   0.4  99.1   0.0    1    763    180    0    0    0  19.46 
20:18:20   0.5   0.0   0.4  99.1   0.0    1    725    147    0    0    0  19.43 
20:18:30   0.5   0.0   0.4  99.1   0.0    1    763    182    3    0   16  19.46 
20:18:40   0.5   0.0   0.5  99.0   0.0    1    722    147    1    0    0  19.43 
20:18:50   0.7   0.0   0.4  98.9   0.0    1    838    177    1    0    1  19.45 
20:19:00   0.5   0.0   0.4  99.1   0.0    1    746    145    1    0    0  19.25 
20:19:10   0.5   0.0   0.4  99.1   0.0    1    746    160    0    0    0  19.49 
20:19:20   0.5   0.0   0.4  99.1   0.0    1    709    147    1    0    0  19.39 
20:19:30   0.6   0.0   0.5  99.0   0.0    1    758    271    3    0    2  19.45 
20:19:40   0.5   0.0   0.5  99.0   0.0    1    803    168   53   33   53  19.45 
20:19:50   0.6   0.0   0.3  99.1   0.0    1    746    162    0    0    0  19.38 
20:20:00   0.6   0.0   0.4  99.0   0.0    1    717    152    0    0    0  19.29 
20:20:10   0.7   0.0   0.5  98.9   0.0    2    771    308    0    0    0  19.31 
20:20:20   0.5   0.0   0.4  99.1   0.0    1    698    144    0    0    0  19.29 
20:20:30   0.5   0.0   0.4  99.1   0.0    1    729    164    2    0    2  19.33 
20:20:40   0.6   0.0   0.4  99.0   0.0    1    698    150    0    0    0  19.17 
20:20:50   0.6   0.0   0.5  98.9   0.0    1    843    184    0    0    0  19.25 
20:21:00   0.6   0.0   0.4  99.0   0.0    1    748    258    0    0    0  19.32 
20:21:10   0.5   0.0   0.4  99.1   0.0    1    711    161    0    0    0  19.47 
20:21:20   0.5   0.0   0.5  99.1   0.0    1    737    158    0    0    0  19.43 
20:21:30   0.5   0.0   0.4  99.1   0.0    1    721    157    2    0    2  19.41 
20:21:40   0.5   0.0   0.4  99.1   0.0    1    713    259    0    0    0  19.53 
20:21:50   0.5   0.0   0.4  99.2   0.0    1    731    165    0    0    0  19.65 
20:22:00   0.5   0.0   0.4  99.1   0.0    1    762    245    1    0    0  19.38 
-------- ----- ----- ----- ----- ----- ---- ------ ------ ---- ---- ---- ------ 
 Average   0.5   0.0   0.4  99.1   0.0  1.1  750.7  190.9  2.4  1.2  2.7  19.39 
  StdDev   0.1   0.0   0.0   0.1   0.0  0.2   41.2   49.8  9.4  5.9  9.8   0.10 
-------- ----- ----- ----- ----- ----- ---- ------ ------ ---- ---- ---- ------ 
 Minimum   0.5   0.0   0.3  98.9   0.0  1.0  698.4  139.0  0.0  0.0  0.0  19.17 
 Maximum   0.7   0.0   0.5  99.2   0.0  2.0  858.3  307.6 53.0 33.0 53.0  19.65 
-------- ----- ----- ----- ----- ----- ---- ------ ------ ---- ---- ---- ------ 
Summary:
System:  19.39 Watts on average with standard deviation 0.10
```

Patched 2.9.x:

```
Running for 300.0 seconds (30 samples at 10.0 second intervals).
Power measurements will start in 180 seconds time.

  Time    User  Nice   Sys  Idle    IO  Run Ctxt/s  IRQ/s Fork Exec Exit  Watts
20:28:30   0.5   0.0   0.4  99.2   0.0    1    751    157    0    0    0  13.61 
20:28:40   0.5   0.0   0.3  99.1   0.0    1    729    150    2    0    2  13.54 
20:28:50   0.5   0.0   0.4  99.1   0.0    1    722    156    0    0    0  13.80 
20:29:00   0.5   0.0   0.4  99.2   0.0    1    716    140    0    0    0  13.64 
20:29:10   0.5   0.0   0.5  99.0   0.0    1    852    419    0    0    0  13.75 
20:29:20   0.5   0.0   0.4  99.1   0.0    1    691    137    2    0    0  13.60 
20:29:30   0.5   0.0   0.4  99.1   0.0    1    722    169    0    0   16  13.51 
20:29:40   0.4   0.0   0.4  99.2   0.0    1    677    133    2    0    2  13.58 
20:29:50   0.6   0.0   0.4  99.0   0.0    1    832    276    2    0    0  13.47 
20:30:00   0.5   0.0   0.3  99.2   0.0    1    700    133    0    0    0  13.55 
20:30:10   0.6   0.0   0.4  99.1   0.0    1    774    160    0    0    0  13.52 
20:30:20   0.8   0.0   0.5  98.7   0.0    1    976    347   55   33   53  13.53 
20:30:30   0.5   0.0   0.4  99.1   0.0    1    763    215    0    0    0  13.75 
20:30:40   0.5   0.0   0.4  99.1   0.0    1    721    153    2    0    2  13.58 
20:30:50   0.7   0.0   0.4  98.9   0.0    1    837    322    0    0    0  13.60 
20:31:00   0.4   0.0   0.4  99.2   0.0    1    698    144    0    0    0  13.34 
20:31:10   0.6   0.0   0.4  99.0   0.0    1    790    313    0    0    0  13.48 
20:31:20   0.5   0.0   0.4  99.1   0.0    1    691    143    0    0    0  13.47 
20:31:30   0.5   0.0   0.4  99.1   0.0    6    718    153    0    0    0  13.39 
20:31:40   0.7   0.0   0.6  98.7   0.0    1    937    554    2    0    2  13.73 
20:31:50   0.6   0.0   0.5  99.0   0.0    1    790    161    0    0    0  13.50 
20:32:00   0.5   0.0   0.4  99.1   0.0    1    734    154    0    0    0  13.66 
20:32:10   0.4   0.0   0.4  99.3   0.0    1    678    132    0    0    0  13.66 
20:32:20   0.5   0.0   0.4  99.1   0.0    1    731    160    0    0    0  13.69 
20:32:30   0.5   0.0   0.4  99.0   0.0    1    806    296    0    0    0  13.66 
20:32:40   0.4   0.0   0.4  99.2   0.0    1    734    158    2    0    2  13.71 
20:32:50   0.4   0.0   0.4  99.2   0.0    1    694    142    0    0    0  13.67 
20:33:00   0.5   0.0   0.4  99.1   0.0    1    751    154    0    0    0  13.62 
20:33:10   0.5   0.0   0.3  99.1   0.0    1    698    139    0    0    0  13.67 
20:33:20   0.5   0.0   0.5  99.0   0.0    1    842    186   53   33   53  13.83 
-------- ----- ----- ----- ----- ----- ---- ------ ------ ---- ---- ---- ------ 
 Average   0.5   0.0   0.4  99.1   0.0  1.2  758.5  201.8  4.1  2.2  4.4  13.60 
  StdDev   0.1   0.0   0.1   0.1   0.0  0.9   72.7   99.1 13.4  8.2 13.3   0.11 
-------- ----- ----- ----- ----- ----- ---- ------ ------ ---- ---- ---- ------ 
 Minimum   0.4   0.0   0.3  98.7   0.0  1.0  676.6  132.2  0.0  0.0  0.0  13.34 
 Maximum   0.8   0.0   0.6  99.3   0.0  6.0  976.2  554.2 55.0 33.0 53.0  13.83 
-------- ----- ----- ----- ----- ----- ---- ------ ------ ---- ---- ---- ------ 
Summary:
System:  13.60 Watts on average with standard deviation 0.11 
```

One curious thing I have found is that changing the value of `pm_qos_latency_tolerance_us` from `25000` to `250000` introduces entries into the `get-feature` table (see output below). I'm not sure if it makes an actual functional difference because any difference between the 2 settings is not significant in powerstat. I guess this is something to take directly to Andy though.

```
get-feature:0x0c (Autonomous Power State Transition), Current value: 0x000001
        Autonomous Power State Transition Enable (APSTE): Enabled
        Auto PST Entries        .................
        Entry[ 0]   
        .................
        Idle Time Prior to Transition (ITPT): 1500 ms
        Idle Transition Power State   (ITPS): 3
        .................
        Entry[ 1]   
        .................
        Idle Time Prior to Transition (ITPT): 1500 ms
        Idle Transition Power State   (ITPS): 3
        .................
        Entry[ 2]   
        .................
        Idle Time Prior to Transition (ITPT): 1500 ms
        Idle Transition Power State   (ITPS): 3
        .................
        Entry[ 3]   
        .................
        Idle Time Prior to Transition (ITPT): 9000 ms
        Idle Transition Power State   (ITPS): 4
        .................
        Entry[ 4]   
        .................
        Idle Time Prior to Transition (ITPT): 0 ms
        Idle Transition Power State   (ITPS): 0
        .................
```

Thanks for making these patches so easily accessible, it's certainly made this experiment so much easier to perform!

**Comment by damige** [comment]
Yes, i think Andy would be a good place to go. I'm glad to hear you appreciate my efforts. 
I will mark this as closed for now. Keep us informed if you talk to Andy about this.

**Comment by hedgepigdaniel** [comment]
I think I worked out why the table is all zeroes - the default maximum total latency for any power state it set to a default of 25ms, but in the Toshiba SSD that both I and @joel-wright use, there is no power state with a total exit latency less than or equal to that. The lowest latency non-operation state has 25ms entry latency and 5ms exit latency for a total of 30ms. The deepest state is 180ms and 170ms for Joel and I respectively.

```
ps    0 : mp:6.00W operational enlat:0 exlat:0 rrt:0 rrl:0
          rwt:0 rwl:0 idle_power:- active_power:-
ps    1 : mp:2.40W operational enlat:0 exlat:0 rrt:1 rrl:1
          rwt:1 rwl:1 idle_power:- active_power:-
ps    2 : mp:1.90W operational enlat:0 exlat:0 rrt:2 rrl:2
          rwt:2 rwl:2 idle_power:- active_power:-
ps    3 : mp:0.0120W non-operational enlat:5000 exlat:25000 rrt:3 rrl:3
          rwt:3 rwl:3 idle_power:- active_power:-
ps    4 : mp:0.0060W non-operational enlat:100000 exlat:70000 rrt:4 rrl:4
          rwt:4 rwl:4 idle_power:- active_power:-
```

If I pass the kernel parameter ``nvme_core.default_ps_max_latency_us=170000` to increase the maximum latency to 170ms. Then the table from `get-feature` has two entries for the two autonomous states.

I did also think before that it was working but maybe I was trying to measure too many things at once. I've done some slightly more scientific tests (powertop for 90s) and confirmed that about 1W is saved with this patch and that kernel parameter compared to having the patch but not setting the kernel parameter.
